src/common/payment.ts

```diff
@@ -207,4 +269,110 @@ export class TurboAuthenticatedPaymentService
       await this.signer.generateSignedRequestHeaders(),
     );
   }
+
+  private async getTargetWalletForFund(): Promise<string> {
+    const { addresses } = await this.httpService.get<TurboInfoResponse>({
+      endpoint: '/info',
+    });
+
+    return addresses[this.token];
+  }
+
+  // TODO: token based backoff
+  private async pollForTxBeingAvailableFromGateway(txId: string) {
+    const maxAttempts = 30;
+    const pollingIntervalMs = 1_000;
+    const initialBackoffMs = 5_000;
+
+    await new Promise((resolve) => setTimeout(resolve, initialBackoffMs));
+
+    let attempts = 0;
+    while (attempts < maxAttempts) {
+      const response = await this.arweave.api.post('/graphql', {
+        query: `
+          query {
+            transaction(id: "${txId}") {
+              recipient
+              owner {
+                address
+              }
+              quantity {
+                winston
+              }
+            }
+          }
+        `,
+      });
+
+      const transaction = response.data.data.transaction;
+
+      if (transaction) {
+        return;
+      }
+      attempts++;
+      this.logger.debug('Transaction not found after polling...', {
+        txId,
+        attempts,
+      });
+      await new Promise((resolve) => setTimeout(resolve, pollingIntervalMs));
+    }
+
+    throw new Error('Transaction not found after polling');
+  }
+
+  public async fund(
+    tokenAmount: BigNumber.Value,
+    feeMultiplier = 1,
+  ): Promise<TurboCryptoFundResponse> {
+    this.logger.debug('Funding account...');
+    const bigNTokenAmount = new BigNumber(tokenAmount);
+
+    const target = await this.getTargetWalletForFund();
+
+    // TODO: token based tx creation
+    const fundTx = await this.arweave.createTransaction({
+      target,
+      quantity: bigNTokenAmount.toString(),
+      data: '',
+    });
+
+    if (feeMultiplier !== 1) {
+      fundTx.reward = BigNumber(fundTx.reward)
+        .times(new BigNumber(feeMultiplier))
+        .toFixed(0, BigNumber.ROUND_UP);
+    }
+
+    const signedTx = await this.signer.signTx(fundTx);
+
+    // TODO: token based tx posting
+    const response = await this.arweave.transactions.post(signedTx);
+
+    if (response.status !== 200) {
+      throw new Error(
+        'Failed to post fund transaction -- ' +
+          `Status ${response.status}, ${response.statusText}`,
+      );
+    }
+
+    this.logger.debug('Posted fund transaction...', { signedTx });
+
+    const txId = signedTx.id;
+
+    // Let transaction settle some time
+    await this.pollForTxBeingAvailableFromGateway(txId);
```

Comment:
see comment above about wrapping this in the try/catch below

Reply:
8bb9b1f